Objects/unionobject.c

```diff
@@ -284,20 +284,37 @@ is_new_type(PyObject *obj)
     return is_typing_module(obj);
 }
 
+// Emulates short-circuiting behavior of the ``||`` operator
+// while also checking negative values.
+#define CHECK_RES(obj, check) { \
+    int result = check(obj); \
```

Comment:
You could pass `check(obj)` instead of `check` and `obj` separately. This would work even if the predicate takes more than one argument.